homeassistant/components/screenlogic/data.py

```diff
@@ -0,0 +1,348 @@
+        lambda flags: EQUIPMENT_FLAG.CHLORINATOR in flags,
+    ),
+}
+
+DEVICE_SUBSCRIPTION = {
+    DEVICE.CONTROLLER: CODE.STATUS_CHANGED,
+    DEVICE.INTELLICHEM: CODE.CHEMISTRY_CHANGED,
+}
+
+
+def get_ha_unit(entity_data: dict) -> StrEnum | str | None:
+    """Return a Home Assistant unit of measurement from a UNIT."""
+    sl_unit = entity_data.get(ATTR.UNIT)
+    return SL_UNIT_TO_HA_UNIT.get(sl_unit, sl_unit)
+
+
+def realize_path_template(
+    template_path: tuple[str | int, ...], data_path: tuple[str | int, ...]
+) -> tuple[str | int, ...]:
+    """Make data path from template and current."""
+    if not data_path or len(data_path) < 3:
+        raise KeyError(
+            f"Missing or invalid required parameter: 'data_path' for template path '{template_path}'"
+        )
+    device, group, data_key = data_path
+    realized_path = []
+    for part in template_path:
+        match part:
+            case PathPart.DEVICE:
+                realized_path.append(device)
+            case PathPart.INDEX | PathPart.KEY:
+                realized_path.append(group)
+            case PathPart.VALUE:
+                realized_path.append(data_key)
+            case _:
+                realized_path.append(part)
+
+    return tuple(realized_path)
+
+
+def process_supported_values(
+    coordinator: ScreenlogicDataUpdateCoordinator,
+    platform_domain: str,
+    supported_devices: SupportedDeviceDescriptions,
+):
+    """Process template data."""
+    entity_registry = er.async_get(coordinator.hass)
+
+    def cleanup_excluded_entity(entity_key: str):
+        """Remove entity if it exists."""
+        assert coordinator.config_entry
+        unique_id = f"{coordinator.config_entry.unique_id}_{entity_key}"
+        if entity_id := entity_registry.async_get_entity_id(
+            platform_domain, SL_DOMAIN, unique_id
+        ):
+            _LOGGER.debug(
+                "Removing existing entity '%s' per data inclusion rule", entity_id
+            )
+            entity_registry.async_remove(entity_id)
+
+    gateway = coordinator.gateway
+
+    for device, device_groups in supported_devices.items():
+        if "*" in device_groups:
+            indexed_values = device_groups.pop("*")
+            for index in gateway.get_data(device):
+                device_groups[index] = indexed_values
+        for group, group_values in device_groups.items():
+            for value_key, value_params in group_values.items():
+                data_path = (device, group, value_key)
+
+                entity_key = generate_unique_id(device, group, value_key)
+
+                if (
+                    inclusion_rule := value_params.get(EntityParameter.INCLUDED)
+                    or DEVICE_INCLUSION_RULES.get(device)
+                ) is not None:
+                    assert isinstance(inclusion_rule, ScreenLogicRule)
+                    if not inclusion_rule.test(gateway, data_path):
+                        cleanup_excluded_entity(entity_key)
```

Comment:
This looks like Home Assistant entity logic

Reply:
This is now in each platform's setup.